### PR TITLE
Change to display sorted list

### PR DESCRIPTION
**Pull request by unblee**
```
$ k8sec list
```

Sorting by using KEY as a key.

# Before

```
NAME                    TYPE                                    KEY             VALUE
dd-agent                Opaque                                  api-key         "*****"
default-token-*****     kubernetes.io/service-account-token     ca.crt          "*****"
default-token-*****     kubernetes.io/service-account-token     namespace       "*****"
default-token-*****     kubernetes.io/service-account-token     token           "*****"
my-docker-registry      kubernetes.io/dockercfg                 .dockercfg      "*****"
```

# After

```
NAME                    TYPE                                    KEY             VALUE
my-docker-registry      kubernetes.io/dockercfg                 .dockercfg      "*****"
dd-agent                Opaque                                  api-key         "*****"
default-token-*****     kubernetes.io/service-account-token     ca.crt          "*****"
default-token-*****     kubernetes.io/service-account-token     namespace       "*****"
default-token-*****     kubernetes.io/service-account-token     token           "*****"
```

### CODE REVIEW

**Comment by dtan4** [comment]
OK, it seems that results are sorted by their **keys** alphabetically.

Now let's check the behavior when there are _multiple_ secret names in the namespace:

```sh-session
$ bin/k8sec list -n sample | grep -v kubernetes.io
NAME                    TYPE                                    KEY             VALUE
app                     Opaque                                  APP_PORT        "80"
app                     Opaque                                  FOO             "baz"
dtan4                   Opaque                                  FOO             "1"
dotenv                  Opaque                                  FOO             "bar"
dtan4                   Opaque                                  GLIDE           "1"
dotenv                  Opaque                                  HOGE            "fuga"
dotenv                  Opaque                                  RAILS_ENV       "production"
app                     Opaque                                  WIDTH           "200"
```

Keys are sorted, but secret names are shuffled. IMO it's a little hard to find what users want.

Would you improve your implementation to print like this?

```sh-session
$ k8sec list -n sample | grep -v kubernetes.io | sort
NAME                    TYPE                                    KEY             VALUE
app                     Opaque                                  APP_PORT        "80"
app                     Opaque                                  FOO             "baz"
app                     Opaque                                  WIDTH           "200"
dotenv                  Opaque                                  FOO             "bar"
dotenv                  Opaque                                  HOGE            "fuga"
dotenv                  Opaque                                  RAILS_ENV       "production"
dtan4                   Opaque                                  FOO             "1"
dtan4                   Opaque                                  GLIDE           "1"
```

**Comment by unblee** [comment]
OK, I'll try.

**Comment by unblee** [comment]
@dtan4 Please review again 🙇

**Comment by dtan4** [comment]
This patches does not cover the case when the secret name is given (e.g., `k8sec list app`). Please check again.

File: cmd/list.go [hunk 1]
```diff
@@ -73,17 +73,32 @@ func doList(cmd *cobra.Command, args []string) error {
...
 		for _, secret := range secrets.Items {
 			for key, value := range secret.Data {
 				if listOpts.base64encode {
 					v = base64.StdEncoding.EncodeToString(value)
 				} else {
 					v = strconv.Quote(string(value))
 				}
-
-				fmt.Fprintln(w, strings.Join([]string{secret.Name, string(secret.Type), key, v}, "\t"))
+				sortedSecrets = append(sortedSecrets, []string{secret.Name, string(secret.Type), key, v})
```
Comment (dtan4) [inline cmd/list.go]: This `sortedSecrets` is NOT sorted at this point. Please consider another name.

Comment (unblee) [inline cmd/list.go]: OK, I'll fix it.

**Comment by unblee** [comment]
Fixed! @dtan4 Please review again 🙇

File: cmd/list.go [hunk 2]
```diff
@@ -80,12 +82,25 @@ func doList(cmd *cobra.Command, args []string) error {
...
+	})
+
+	// sort by NAME
+	sort.SliceStable(formattedSecrets, func(i, j int) bool {
+		return formattedSecrets[i][0] < formattedSecrets[j][0]
+	})
+
+	for _, secret := range formattedSecrets {
+		fmt.Fprintln(w, strings.Join(secret, "\t"))
+	}
```
Comment (dtan4) [inline cmd/list.go]: Response of `ListSecrets` is already sorted alphabetically (see https://github.com/kubernetes/kubernetes/blob/v1.6.6/staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_helper.go#L456-L460 how k8s retrieves objects).
So we don't have to sort by key again!

Why don't you sort and print key-values in `for _, secret := range secrets.Items { ... }` loop?

Comment (unblee) [inline cmd/list.go]: Because I didn't want to duplicate the "sort" and "print" processes.
I'll fix it.

Comment (dtan4) [inline cmd/list.go]: you're right, print process can be leave as it is to keep DRY code.

**Comment by unblee** [comment]
Fixed!! @dtan4 Please review again 🙇 🙇

**Comment by unblee** [comment]
Sorry, Please wait 🙇

**Comment by unblee** [comment]
@dtan4 Please review again 🙇 🙇 🙇

**Comment by unblee** [comment]
@dtan4 Please review again 🙇

File: cmd/list.go [hunk 3]
```diff
@@ -65,27 +75,65 @@ func doList(cmd *cobra.Command, args []string) error {
...
+					Key:   kv.k,
+					Value: kv.v,
+				})
 			}
 		}
 	}
 
+	for _, secret := range secretList {
+		a := []string{secret.Name, secret.Type, secret.Key, secret.Value}
+		fmt.Fprintln(w, strings.Join(a, "\t"))
```
[the remark this answers is not in the document]
Comment (unblee) [reply cmd/list.go]: OK

[the remark this answers is not in the document]
Comment (unblee) [reply cmd/list.go]: Fixed.

File: cmd/list.go [hunk 4]
```diff
@@ -52,6 +60,8 @@ func doList(cmd *cobra.Command, args []string) error {
 
 	var v string
 
+	secretList := []Secret{}
```
Comment (dtan4) [inline cmd/list.go]: It seems most of Go code use the plural form to indicate slice, like `secrets`.

How about renaming this to `secrets`, and rename _existing_ `secrets` (the result of `k8sclient.ListSecrets()`) to `ss` ?
IMO it's easy-to understand the variable scope.

Comment (unblee) [reply cmd/list.go]: Fixed.

**Comment by dtan4** [comment]
LGTM 👍

**Comment by unblee** [comment]
Thanks!!!

**Comment by creasty** [comment]
cool :+1:

**Pull request merged**